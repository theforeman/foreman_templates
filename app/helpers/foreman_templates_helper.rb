module ForemanTemplatesHelper
  def edit_paths
    Template.subclasses.reduce({}) do |memo, subclass|
      memo.tap do |acc|
        path_part = "#{subclass.name.underscore.pluralize}_path"
        acc[subclass.name] = "#{Rails.application.routes.url_helpers.public_send(path_part)}/:id/edit"
      end
    end
  end
end
