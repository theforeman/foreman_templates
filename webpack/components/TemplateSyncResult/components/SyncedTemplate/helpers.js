import React from 'react';
import { isEmpty } from 'lodash';
import { Icon } from 'patternfly-react';

import IconInfoItem from './IconInfoItem';
import EmptyInfoItem from './EmptyInfoItem';
import StringInfoItem from './StringInfoItem';

export const itemIteratorId = (template, attr) => `${template.name}-${attr}`;

export const additionalInfo = template => {
  const infoAttrs = [
    'locked',
    'snippet',
    'humanizedClassName',
    'kind',
    'templateFile',
  ];

  return infoAttrs.map(attr => {
    const key = itemIteratorId(template, attr);

    const classNameMap = { Ptable: 'Partition Table' };

    if (!template[attr]) {
      return <EmptyInfoItem template={template} attr={attr} key={key} />;
    }

    switch (attr) {
      case 'locked':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="lock"
            tooltipText="Locked"
            key={key}
          />
        );
      case 'snippet':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="check"
            tooltipText="Snippet"
            key={key}
          />
        );
      case 'humanizedClassName':
        return (
          <StringInfoItem
            template={template}
            attr={attr}
            translate
            mapAttr={(templateObj, attribute) =>
              classNameMap[templateObj[attribute]]
                ? classNameMap[templateObj[attribute]]
                : templateObj[attribute]
            }
            key={key}
          />
        );
      case 'kind':
        return <StringInfoItem template={template} attr={attr} key={key} />;
      case 'templateFile':
        return (
          <StringInfoItem template={template} attr={attr} key={key} elipsed />
        );
      default:
        return '';
    }
  });
};

export const itemLeftContentIcon = template => {
  const iconName = isEmpty(aggregatedErrors(template))
    ? 'ok'
    : 'error-circle-o';
  return <Icon name={iconName} size="sm" type="pf" />;
};

export const templateHeading = (template, editPath) => {
  if (template.id && template.canEdit) {
    return (
      <a
        href={editPath.replace(':id', template.id)}
        target="_blank"
        rel="noopener noreferrer"
      >
        {template.name}
      </a>
    );
  }
  return template.name || ' ';
};

export const templateErrors = template => {
  if (Object.keys(aggregatedErrors(template)).length !== 0) {
    const res = Object.keys(aggregatedErrors(template)).map(key => (
      <li key={itemIteratorId(template, key)}>
        {formatError(key, aggregatedErrors(template)[key])}
      </li>
    ));
    return <ul>{res}</ul>;
  }
  return <span>There were no errors.</span>;
};

const aggregatedErrors = template => {
  const err = { ...template.errors } || {};
  if (template.additional_errors) {
    err.additional = template.additional_errors;
  }

  return err;
};

const formatError = (key, value) => {
  const omitKeys = ['base', 'additional'];
  if (omitKeys.reduce((memo, item) => memo || key === item, false)) {
    return value;
  }

  return `${key}: ${value}`;
};
