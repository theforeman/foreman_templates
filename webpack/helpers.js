import React from 'react';

import {
  propsToCamelCase,
} from 'foremanReact/common/helpers';

export const deepPropsToCamelCase = obj => {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }
  if (Array.isArray(obj)) {
    return obj.map(deepPropsToCamelCase);
  }
  const transformed = propsToCamelCase(obj);
  return Object.keys(transformed).reduce((memo, key) => {
    memo[key] = deepPropsToCamelCase(transformed[key]);
    return memo;
  }, {})
};

// TODO: extract to core
export const withProtectedView = (
  ProtectedComponent,
  ProtectionComponent,
  protectionFn,
  extraProtectionProps = {}
) => props =>
  protectionFn(props) ? (
    <ProtectedComponent {...props} />
  ) : (
    <ProtectionComponent {...props} {...extraProtectionProps}/>
  );

export default withProtectedView;
