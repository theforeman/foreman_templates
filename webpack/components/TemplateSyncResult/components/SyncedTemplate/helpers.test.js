import { testSelectorsSnapshotWithFixtures } from 'react-redux-test-utils';

import { expandableContent } from './helpers';

const fixtures = {
  'should return no errors': () => expandableContent({}),
  'should return additional info': () =>
    expandableContent({ additionalInfo: 'This is additional info' }),
  'should return additional error': () =>
    expandableContent({ additionalErrors: 'These are additional errors' }),
  'should return attribute errors': () =>
    expandableContent({
      errors: {
        base: ['could not be processed'],
        name: ["can't be blank", 'has invalid format'],
      },
    }),
};

describe('expandableContent', () =>
  testSelectorsSnapshotWithFixtures(fixtures));
