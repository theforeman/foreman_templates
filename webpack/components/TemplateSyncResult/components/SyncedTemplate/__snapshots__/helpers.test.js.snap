// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandableContent should return additional error 1`] = `
<ul>
  <li>
    These are additional errors
  </li>
</ul>
`;

exports[`expandableContent should return additional info 1`] = `
<ul>
  <li>
    This is additional info
  </li>
</ul>
`;

exports[`expandableContent should return attribute errors 1`] = `
<ul>
  <li>
    could not be processed
  </li>
  <li>
    name: can't be blank
  </li>
  <li>
    name: has invalid format
  </li>
</ul>
`;

exports[`expandableContent should return no errors 1`] = `
<span>
  There were no errors.
</span>
`;
